---
title: "01-Reformat_to_SummarizedExperiment"
author: "Valentin Marteau"
params:
  data: "../data"
  results:  "../results"
  lib:  "../lib"
  maxcores: 6
output:
  html_notebook:
    theme: spacelab
    highlight: textmate
    toc: yes
    number_sections: true
    toc_depth: 3
    toc_float: true
---

```{r, results = "hide"}
# Load required packages
library(SummarizedExperiment)
library(tidyverse)
library(rtracklayer)
```

# Reformat and check data and metadata
```{r}
# Read raw data into memory
raw_data <- read_tsv(file = file.path(params$data, "salmon.merged.gene_counts.tsv"))
metadata <- read_csv2(file = file.path(params$data, "metadata.csv"))

# Check metadata, convert to factors (but keep covariates as numerical) and abbreviate labels
metadata  <- metadata |> mutate_at(vars(-c("seq_ID", "week", "day", "GW", "weight")), factor)
metadata$place <- recode_factor(metadata$place, "control" = "Ctrl", "control_mus" = "CtrlMus")

# Correct typo (upper/lower case "Sectio")
metadata$sectio <- recode_factor(metadata$sectio, "Prim. Sectio" = "Prim", "Sek. Sectio" = "Sek", "Prim. sectio" = "Prim")

# Simplify sample names for both data and metadata (need to be equal!)
remove <- c("AK19_", "[[:punct:]]")
metadata$sample <- metadata$seq_ID |> str_remove_all(paste(remove, collapse = "|"))
metadata <- column_to_rownames(metadata, "sample")

# Remove ENSG version from gene id
remove_ensg_version <- function(x) gsub("\\.[0-9]*$", "", x)
raw_data$gene_id <- remove_ensg_version(raw_data$gene_id)

# Generate count matrix
count_mat <- as.matrix(ceiling(raw_data[, 3:23]))
rownames(count_mat) <- raw_data$gene_id
colnames(count_mat) <- rownames(metadata)

# Reorder Samples in ascending order
ord <- order(as.numeric(str_remove_all(rownames(metadata), "TG")))
count_mat <- count_mat[, ord]
metadata <- metadata[ord, ]
```
# Compile SummarizedExperiment object
The gtf/gff file can also be downloaded at https://www.gencodegenes.org/human/. Just need to make sure that it is the same version as the gtf file that was used to generate the counts data (In this case for example version 38)! When getting the file from the internet you might need to synchronize the gene_ids between the counts data and the GRangesList object (Need to be the same when compiling the SummarizedExperiment object).
```{r}
# Get gtf file that was used to generate counts data (path on zeus) and import as GRanges (Genomic Ranges) object
#file = file.path("/data/genomes/hg38/annotation/gencode/gencode.v38.primary_assembly.annotation.gtf")
#gtf <- rtracklayer::import(file)

gtf <- readRDS(file = file.path(params$data, "PAS_gtf.rds"))

# Remove ENSG version from gene id and generate and split by gene id to generate GRangesList
mcols(gtf)$gene_id <- remove_ensg_version(mcols(gtf)$gene_id)
grl <- split(gtf, mcols(gtf)[5])

# Create SummarizedExperiment object
se <- SummarizedExperiment(assays = count_mat,
                           rowRanges = grl,
                           colData = metadata)

# Remove all genes with no counts at all
ridx <- rowMeans(assay(se)) > 0
se <- se[ridx, , drop = FALSE]

# Rename first assay
names(assays(se))[1] <- "raw_counts"

# Add experimental description
metadata(se)$Project_info <- "Placenta Accreta Spectrum (PAS) bulk RNAseq data"
```
```{r}
saveRDS(se, file.path(params$results, "PAS_SummarizedExperiment"))
```





